### PR TITLE
Move code to assign tax information into shared parent

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Code cleanup around tax assignments

Before
----------------------------------------
Code not in a shareable place

After
----------------------------------------
Code more shareable - 3 other classes need to share this

Technical Details
----------------------------------------
@mattwire this is a bit of a chunkier cleanup so hopefully it's not too much. I was going to just clarify the tpl logic (ie let's specify whether to displayLineItems rather than intuiting if from a combination of parameters that are not directly related & then always assign line items) but then I went a bit further in cleaning up the form function. I wrote more comments :-)

Comments
----------------------------------------

### CODE REVIEW

File: CRM/Financial/Form/FrontEndPaymentFormTrait.php [hunk 1]
```diff
@@ -43,25 +43,39 @@ trait CRM_Financial_Form_FrontEndPaymentFormTrait {
    * CRM_Invoicing_Utils class with a potential end goal of moving this handling to an extension.
    *
    * @param $tplLineItems
-   *
-   * @return array
    */
-  protected function alterLineItemsForTemplate($tplLineItems) {
-    $getTaxDetails = FALSE;
+  protected function alterLineItemsForTemplate(&$tplLineItems) {
+    if (CRM_Invoicing_Utils::isInvoicingEnabled()) {
```
Comment (seamuslee001) [inline CRM/Financial/Form/FrontEndPaymentFormTrait.php]: @eileenmcnaughton should there be ! before the CRM...?

Comment (eileenmcnaughton) [reply CRM/Financial/Form/FrontEndPaymentFormTrait.php]: opps - fixed

**Comment by seamuslee001** [comment]
@eileenmcnaughton need some jenkins love https://test.civicrm.org/job/CiviCRM-Core-PR/25500/checkstyleResult/new/

**Comment by eileenmcnaughton** [comment]
@seamuslee001 did we change the standard?
Those lines are not in the PR & seem 'ok' to me

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
Reply to myself - we obviously HAVE changed the standard - I see it hit @agilewarealok as well today. Hopefully that won't be too painful

File: CRM/Financial/Form/FrontEndPaymentFormTrait.php [hunk 2]
```diff
@@ -43,25 +43,39 @@ trait CRM_Financial_Form_FrontEndPaymentFormTrait {
...
+    // them but not be able to 'format them right' because they are
+    // potentially being used for 'something else'.
+    // @todo invoicing code - please feel the hate. Also move this 'hook-like-bit'
+    // to the CRM_Invoicing_Utils class.
     foreach ($tplLineItems as $key => $value) {
       foreach ($value as $k => $v) {
         if (isset($v['tax_rate']) && $v['tax_rate'] != '') {
-          $getTaxDetails = TRUE;
+          $this->assign('getTaxDetails', TRUE);
+          $this->assign('taxTerm', CRM_Invoicing_Utils::getTaxTerm());
```
Comment (mattwire) [inline CRM/Financial/Form/FrontEndPaymentFormTrait.php]: We are calling these lines for every lineitem when they only need assigning once.  I can see why, just not sure if that's really a good idea.

File: CRM/Financial/Form/FrontEndPaymentFormTrait.php [hunk 3]
```diff
@@ -43,25 +43,39 @@ trait CRM_Financial_Form_FrontEndPaymentFormTrait {
...
+  /**
+   * Assign line items to the template.
+   *
+   * @param $tplLineItems
+   */
+  protected function assignLineItemsToTemplate($tplLineItems) {
+    // @todo this should be a hook that invoicing code hooks into rather than a call to it.
+    $this->alterLineItemsForTemplate($tplLineItems);
+    $this->assign('lineItem', $tplLineItems);
   }
```
Comment (mattwire) [inline CRM/Financial/Form/FrontEndPaymentFormTrait.php]: ok

File: templates/CRM/Contribute/Form/Contribution/Confirm.tpl [hunk 4]
```diff
@@ -44,16 +44,16 @@
 
   {include file="CRM/Contribute/Form/Contribution/MembershipBlock.tpl" context="confirmContribution"}
 
-  {if $amount GTE 0 OR $minimum_fee GTE 0 OR ( $priceSetID and $lineItem ) }
+  {if $amount GTE 0 OR $minimum_fee GTE 0 OR ( $isDisplayLineItems and $lineItem ) }
```
Comment (mattwire) [inline templates/CRM/Contribute/Form/Contribution/Confirm.tpl]: ok

File: templates/CRM/Contribute/Form/Contribution/Confirm.tpl [hunk 5]
```diff
@@ -44,16 +44,16 @@
...
     <div class="crm-group amount_display-group">
       {if !$useForMember}
         <div class="header-dark">
-          {if !$membershipBlock AND $amount OR ( $priceSetID and $lineItem ) }{ts}Contribution Amount{/ts}{else}{ts}Membership Fee{/ts} {/if}
+          {if !$membershipBlock AND $amount OR ( $isDisplayLineItems and $lineItem ) }{ts}Contribution Amount{/ts}{else}{ts}Membership Fee{/ts} {/if}
```
Comment (mattwire) [inline templates/CRM/Contribute/Form/Contribution/Confirm.tpl]: ok

File: templates/CRM/Contribute/Form/Contribution/Confirm.tpl [hunk 6]
```diff
@@ -44,16 +44,16 @@
...
         </div>
       {/if}
       <div class="display-block">
         {if !$useForMember}
-          {if $lineItem and $priceSetID}
+          {if $lineItem and $isDisplayLineItems}
```
Comment (mattwire) [inline templates/CRM/Contribute/Form/Contribution/Confirm.tpl]: ok

File: CRM/Contribute/Form/Contribution/Confirm.php [hunk 7]
```diff
@@ -1811,8 +1816,7 @@ protected function getIsPending() {
     // The concept of contributeMode is deprecated.
     // the is_monetary concept probably should be too as it can be calculated from
     // the existence of 'amount' & seems fragile.
-    if (((isset($this->_contributeMode)) || !empty
-        ($this->_params['is_pay_later'])
+    if (((isset($this->_contributeMode)) || !empty($this->_params['is_pay_later'])
```
Comment (mattwire) [inline CRM/Contribute/Form/Contribution/Confirm.php]: ok

File: CRM/Contribute/Form/Contribution/Confirm.php [hunk 8]
```diff
@@ -511,15 +511,20 @@ public function buildQuickForm() {
...
-    else {
+    $this->assignLineItemsToTemplate($tplLineItems);
+
+    $isDisplayLineItems = $this->_priceSetId && !CRM_Core_DAO::getFieldValue('CRM_Price_DAO_PriceSet', $this->_priceSetId, 'is_quick_config');
+
+    $this->assign('isDisplayLineItems', $isDisplayLineItems);
+
+    if (!$isDisplayLineItems) {
+      // quickConfig is deprecated in favour of isDisplayLineItems. Lots of logic has been harnessed to quick config
+      // whereas isDisplayLineItems is specific & clear.
```
Comment (mattwire) [inline CRM/Contribute/Form/Contribution/Confirm.php]: ok

File: CRM/Contribute/Form/Contribution/Confirm.php [hunk 9]
```diff
@@ -511,15 +511,20 @@ public function buildQuickForm() {
     // Make a copy of line items array to use for display only
     $tplLineItems = $this->_lineItem;
     if (CRM_Invoicing_Utils::isInvoicingEnabled()) {
-      list($getTaxDetails, $tplLineItems) = $this->alterLineItemsForTemplate($tplLineItems);
-      $this->assign('getTaxDetails', $getTaxDetails);
-      $this->assign('taxTerm', CRM_Invoicing_Utils::getTaxTerm());
+      // @todo $params seems like exactly the wrong place to get totalTaxAmount from
+      // this is a calculated variable so we it should be transparent how we
+      // calculated it rather than coming from 'params'
```
Comment (mattwire) [inline CRM/Contribute/Form/Contribution/Confirm.php]: ok

File: CRM/Contribute/Form/Contribution/Confirm.php [hunk 10]
```diff
@@ -412,7 +412,7 @@ public function preProcess() {
             }
 
             $isPrimary = 1;
-            if (isset ($this->_params['onbehalf_location'][$blockName])
+            if (isset($this->_params['onbehalf_location'][$blockName])
```
Comment (mattwire) [inline CRM/Contribute/Form/Contribution/Confirm.php]: ok

**Review by mattwire — CHANGES_REQUESTED** [review]

File: CRM/Financial/Form/FrontEndPaymentFormTrait.php (hunk 2 above)
Comment (JoeMurray) [reply CRM/Financial/Form/FrontEndPaymentFormTrait.php]: Not sure I agree. It should be possible for different line items to have different taxes configured, leading to a different taxTerm being displayed for different line items.

Comment (mattwire) [reply CRM/Financial/Form/FrontEndPaymentFormTrait.php]: @JoeMurray Agree.  But `getTaxDetails` and `taxTerm` are being assigned globally so they'll just overwrite the same var each time the loop is run.  `tax_rate` *is* per lineitem, but `taxTerm` is a global setting under "Tax and Invoicing" and CiviCRM doesn't currently support multiple tax terms (eg. Sales Tax, IVA, VAT).

Comment (eileenmcnaughton) [reply CRM/Financial/Form/FrontEndPaymentFormTrait.php]: @mattwire yeah they only need to be assigned once but it seemed really 'cheap' & to make the code more readable. It doesn't change functionality at all

Comment (mattwire) [reply CRM/Financial/Form/FrontEndPaymentFormTrait.php]: Ok, still not completely comfortable as I worry it has the potential to introduce performance issues later (eg. if we made getTaxTerm() do something wierd and wonderful).  If you add a one line comment `// These only need assigning once, but code is more readable with them here` then I'll merge

**Comment by eileenmcnaughton** [comment]
done

**Review by mattwire — APPROVED** [review]

**Comment by eileenmcnaughton** [comment]
merging as tests passed (I did just rebase since another commit had snuck in but that's in master & not relevant to test success)

**Pull request merged**